Fix -Wreturn-stack-address warnings

This may be an upstream bug, not sure.

Looks like in two files (MSFTPasses.cpp), (Reduction.cpp) we attempt to use the casting infrastructure in a way that Clang complains about:

`MSFTPasses.cpp`:
```
[2/8] Building CXX object lib/Dialect/MSFT/CMakeFiles/obj.CIRCTMSFT.dir/MSFTPasses.cpp.o
In file included from /home/will/src/sifive/circt/lib/Dialect/MSFT/MSFTPasses.cpp:9:
In file included from /home/will/src/sifive/circt/include/circt/Dialect/MSFT/MSFTPasses.h:12:
In file included from /home/will/src/sifive/circt/include/circt/Dialect/MSFT/MSFTOps.h:16:
In file included from /home/will/src/sifive/circt/include/circt/Dialect/HW/HWAttributes.h:12:
In file included from /home/will/src/sifive/circt/llvm/mlir/include/mlir/IR/Attributes.h:12:
In file included from /home/will/src/sifive/circt/llvm/mlir/include/mlir/IR/AttributeSupport.h:16:
In file included from /home/will/src/sifive/circt/llvm/mlir/include/mlir/IR/MLIRContext.h:12:
In file included from /home/will/src/sifive/circt/llvm/mlir/include/mlir/Support/LLVM.h:24:
/home/will/src/sifive/circt/llvm/llvm/include/llvm/Support/Casting.h:491:69: warning: returning reference to local temporary object [-Wreturn-stack-address]
  static inline CastReturnType castFailed() { return CastReturnType(nullptr); }
                                                                    ^~~~~~~
/home/will/src/sifive/circt/llvm/llvm/include/llvm/Support/Casting.h:495:14: note: in instantiation of member function 'llvm::CastInfo<circt::msft::InstanceOp, const circt::hw::HWInstanceLike>::castFailed' requested here
      return castFailed();
             ^
/home/will/src/sifive/circt/llvm/llvm/include/llvm/Support/Casting.h:598:36: note: in instantiation of member function 'llvm::CastInfo<circt::msft::InstanceOp, const circt::hw::HWInstanceLike>::doCastIfPossible' requested here
  return CastInfo<To, const From>::doCastIfPossible(Val);
                                   ^
1 warning generated.
```

`Reduction.cpp`:
```
[1/8] Building CXX object tools/circt-reduce/CMakeFiles/circt-reduce.dir/Reduction.cpp.o
In file included from /home/will/src/sifive/circt/tools/circt-reduce/Reduction.cpp:13:
In file included from /home/will/src/sifive/circt/tools/circt-reduce/Reduction.h:19:
In file included from /home/will/src/sifive/circt/llvm/mlir/include/mlir/IR/BuiltinOps.h:16:
In file included from /home/will/src/sifive/circt/llvm/mlir/include/mlir/IR/OpImplementation.h:16:
In file included from /home/will/src/sifive/circt/llvm/mlir/include/mlir/IR/BuiltinTypes.h:12:
In file included from /home/will/src/sifive/circt/llvm/mlir/include/mlir/IR/BuiltinAttributeInterfaces.h:12:
In file included from /home/will/src/sifive/circt/llvm/mlir/include/mlir/IR/AffineMap.h:17:
In file included from /home/will/src/sifive/circt/llvm/mlir/include/mlir/IR/AffineExpr.h:17:
In file included from /home/will/src/sifive/circt/llvm/mlir/include/mlir/Support/LLVM.h:24:
/home/will/src/sifive/circt/llvm/llvm/include/llvm/Support/Casting.h:491:69: warning: returning reference to local temporary object [-Wreturn-stack-address]
  static inline CastReturnType castFailed() { return CastReturnType(nullptr); }
                                                                    ^~~~~~~
/home/will/src/sifive/circt/llvm/llvm/include/llvm/Support/Casting.h:495:14: note: in instantiation of member function 'llvm::CastInfo<circt::firrtl::FModuleOp, const circt::firrtl::FModuleLike>::castFailed' requested here
      return castFailed();
             ^
/home/will/src/sifive/circt/llvm/llvm/include/llvm/Support/Casting.h:598:36: note: in instantiation of member function 'llvm::CastInfo<circt::firrtl::FModuleOp, const circt::firrtl::FModuleLike>::doCastIfPossible' requested here
  return CastInfo<To, const From>::doCastIfPossible(Val);
                                   ^
1 warning generated.
```

These warnings aren't as helpful as one might hope for.

--- a/lib/Dialect/MSFT/MSFTPasses.cpp
+++ b/lib/Dialect/MSFT/MSFTPasses.cpp
@@ -1568,7 +1568,8 @@ void MSFTPassCommon::dedupInputs(MSFTModuleOp mod) {
   // loopbacks for all the instances.
   if (instantiations.size() != 1)
     return;
-  InstanceOp inst = dyn_cast<InstanceOp>(instantiations[0]);
+  InstanceOp inst =
+      dyn_cast<InstanceOp>(static_cast<Operation *>(instantiations[0]));
   if (!inst)
     return;
 
@@ -1619,7 +1620,8 @@ void MSFTPassCommon::sinkWiresDown(MSFTModuleOp mod) {
   // loopbacks for all the instances.
   if (instantiations.size() != 1)
     return;
-  InstanceOp inst = dyn_cast<InstanceOp>(instantiations[0]);
+  InstanceOp inst =
+      dyn_cast<InstanceOp>(static_cast<Operation *>(instantiations[0]));
   if (!inst)
     return;
 

--- a/tools/circt-reduce/Reduction.cpp
+++ b/tools/circt-reduce/Reduction.cpp
@@ -72,7 +72,8 @@ static llvm::Optional<firrtl::FModuleOp>
 findInstantiatedModule(firrtl::InstanceOp instOp, SymbolCache &symbols) {
   auto *tableOp = SymbolTable::getNearestSymbolTable(instOp);
   auto moduleOp = dyn_cast<firrtl::FModuleOp>(
-      instOp.getReferencedModule(symbols.getSymbolTable(tableOp)));
+      instOp.getReferencedModule(symbols.getSymbolTable(tableOp))
+          .getOperation());
   return moduleOp ? llvm::Optional(moduleOp)
                   : llvm::Optional<firrtl::FModuleOp>();
 }
@@ -1031,7 +1032,7 @@ struct EagerInliner : public Reduction {
       return 0;
     auto tableOp = SymbolTable::getNearestSymbolTable(instOp);
     auto moduleOp = instOp.getReferencedModule(symbols.getSymbolTable(tableOp));
-    if (!isa<firrtl::FModuleOp>(moduleOp))
+    if (!isa<firrtl::FModuleOp>(moduleOp.getOperation()))
       return 0;
     return symbols.getSymbolUserMap(tableOp).getUsers(moduleOp).size() == 1;
   }
@@ -1054,7 +1055,8 @@ struct EagerInliner : public Reduction {
     }
     auto tableOp = SymbolTable::getNearestSymbolTable(instOp);
     auto moduleOp = cast<firrtl::FModuleOp>(
-        instOp.getReferencedModule(symbols.getSymbolTable(tableOp)));
+        instOp.getReferencedModule(symbols.getSymbolTable(tableOp))
+            .getOperation());
     for (auto &op : llvm::make_early_inc_range(*moduleOp.getBodyBlock())) {
       op.remove();
       builder.insert(&op);

